From 97cf4dc660c6763c104b51130ab510077c0e3fde Mon Sep 17 00:00:00 2001
From: zachary <zhangzg@marvell.com>
Date: Tue, 28 Feb 2017 00:11:52 +0800
Subject: [PATCH 0716/1200] phy: marvell: a3700: Make USB compatibility of
 device and host mode.

- On A3700, there are both USB host and device controller,
  but on PHY level the configuration is the same.
- The patch adopts PHY_TYPE_USB3 to support both
  host mode and device mode simultaneously.

Change-Id: I2c9eb301025831c608ac7d09e907c813166f4da0
Signed-off-by: zachary <zhangzg@marvell.com>
Signed-off-by: Ken Ma <make@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/53605
Reviewed-by: Igal Liberman <igall@marvell.com>
Tested-by: Igal Liberman <igall@marvell.com>
---
 drivers/phy/marvell/comphy_a3700.c | 27 +++++++--------------------
 drivers/phy/marvell/comphy_a3700.h |  2 ++
 2 files changed, 9 insertions(+), 20 deletions(-)

diff --git a/drivers/phy/marvell/comphy_a3700.c b/drivers/phy/marvell/comphy_a3700.c
index d42b05ceb5..f93ff6c8fe 100644
--- a/drivers/phy/marvell/comphy_a3700.c
+++ b/drivers/phy/marvell/comphy_a3700.c
@@ -327,7 +327,7 @@ static int comphy_sata_power_up(u32 invert)
  *
  * return: 1 if PLL locked (OK), 0 otherwise (FAIL)
  */
-static int comphy_usb3_power_up(u32 type, u32 speed, u32 invert)
+static int comphy_usb3_power_up(u32 speed, u32 invert)
 {
 	int	ret;
 
@@ -457,24 +457,11 @@ static int comphy_usb3_power_up(u32 type, u32 speed, u32 invert)
 		printf("Failed to lock USB3 PLL\n");
 
 	/*
-	 * Set Soft ID for Host mode (Device mode works with Hard ID
-	 * detection)
+	 * No matter host mode and device mode, it works with Hard ID detection.
+	 * Unset DP and DM pulldown for USB2 Device mode
 	 */
-	if (type == COMPHY_TYPE_USB3_HOST0) {
-		/*
-		 * set   BIT0: set ID_MODE of Host/Device = "Soft ID" (BIT1)
-		 * clear BIT1: set SOFT_ID = Host
-		 * set   BIT4: set INT_MODE = ID. Interrupt Mode: enable
-		 *             interrupt by ID instead of using both interrupts
-		 *             of HOST and Device ORed simultaneously
-		 *             INT_MODE=ID in order to avoid unexpected
-		 *             behaviour or both interrupts together
-		 */
-		reg_set((void __iomem *)USB32_CTRL_BASE,
-			usb32_ctrl_id_mode | usb32_ctrl_int_mode,
-			usb32_ctrl_id_mode | usb32_ctrl_soft_id |
-			usb32_ctrl_int_mode);
-	}
+	reg_set((void __iomem *)USB2_OTG_PHY_CTRL_ADDR, 0x0,
+		rb_usb2_dp_pulldn_dev_mode | rb_usb2_dm_pulldn_dev_mode);
 
 	/* Disbale VBus interrupt which will be enable again in kernel */
 	reg_set((void __iomem *)USB3_TOP_INT_ENABLE_REG, 0x0, vbus_int_enable);
@@ -954,10 +941,10 @@ int comphy_a3700_init(struct chip_serdes_phy_config *chip_cfg,
 						   comphy_map->invert);
 			break;
 
+		case COMPHY_TYPE_USB3:
 		case COMPHY_TYPE_USB3_HOST0:
 		case COMPHY_TYPE_USB3_DEVICE:
-			ret = comphy_usb3_power_up(comphy_map->type,
-						   comphy_map->speed,
+			ret = comphy_usb3_power_up(comphy_map->speed,
 						   comphy_map->invert);
 			break;
 
diff --git a/drivers/phy/marvell/comphy_a3700.h b/drivers/phy/marvell/comphy_a3700.h
index 8fb6795f09..7c7bd329e0 100644
--- a/drivers/phy/marvell/comphy_a3700.h
+++ b/drivers/phy/marvell/comphy_a3700.h
@@ -189,6 +189,8 @@
 #define USB2_OTG_PHY_CTRL_ADDR		(0x820 + USB2PHY_BASE)
 #define rb_usb2phy_suspm		BIT(14)
 #define rb_usb2phy_pu			BIT(0)
+#define rb_usb2_dp_pulldn_dev_mode	BIT(5)
+#define rb_usb2_dm_pulldn_dev_mode	BIT(6)
 
 #define USB2_PHY_OTG_CTRL_ADDR		(0x34 + USB2PHY_BASE)
 #define rb_pu_otg			BIT(4)
-- 
2.22.0

